### PR TITLE
Extract image pipeline into external service

### PR DESCRIPTION
**Pull request by StealWonders**
We use imaginary to resize and convert the images. The GraphQL API only has to validate and save the images into a shared volume

### CODE REVIEW

File: internal/graphql/resolvers/helper.go [hunk 1]
```diff
@@ -3,24 +3,20 @@ package resolvers
...
 )
 
-func (r *Resolver) transcodeAndStoreImage(ctx context.Context, image *models.ImageInput) (string, error) {
-	if image.Image.Size > r.ImageProcessor.GetMaxImageSize() {
-		return "", errors.New("image size is too big")
-	}
-
-	imageBytes, err := ioutil.ReadAll(image.Image.File)
+func (r *Resolver) storeImage(ctx context.Context, image *models.ImageInput) (string, error) {
+	imageBytes, err := io.ReadAll(image.Image.File)
```
Comment (tobiges) [inline internal/graphql/resolvers/helper.go:14]: why remove the size check? but I guess its already covered by the gqlgen pload limit.

File: internal/images/images.go [hunk 2]
```diff
@@ -1,65 +1,45 @@
...
-			path, err = ip.GetResized(imageStoreID, width, height)
-			if err == imageprocessor.ErrOriginalImageNotFound {
-				c.AbortWithError(http.StatusNotFound, err)
-				return
-			} else if err != nil {
-				c.AbortWithError(http.StatusInternalServerError, err)
-				return
-			}
+		path := ip.GetImagePath(imageHash)
+		print(path)
```
Comment (tobiges) [inline internal/images/images.go:33]: remove the prints

File: internal/images/images.go [hunk 3]
```diff
@@ -1,65 +1,45 @@
...
 
 type ImageParams struct {
-	ImageProcessor *imageprocessor.ImageProcessor
+	ImageUploader *imageuploader.ImageUploader
 }
 
 func Run(g *gin.RouterGroup, params *ImageParams) error {
-	relativePath := "/:imageStoreID"
-	handler := imageHandler(params.ImageProcessor)
+	relativePath := "/:imageHash"
```
Comment (tobiges) [inline internal/images/images.go:15]: I preferred the old name, but its ok

File: internal/images/images.go [hunk 4]
```diff
@@ -1,65 +1,45 @@
...
-				return
-			} else if err != nil {
-				c.AbortWithError(http.StatusInternalServerError, err)
-				return
-			}
+		path := ip.GetImagePath(imageHash)
+		print(path)
+
+		// check if file exists and send it if it does
+		if _, err := os.Stat(path); !os.IsNotExist(err) {
```
Comment (tobiges) [inline internal/images/images.go:36]: I think the other error cases beside IsNotExist should also be handeled

File: pkg/imageuploader/imageuploader.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	return iu.maxImageSize
+}
+
+func IsImageHashValid(imageHash string) bool {
+	return isValidImageHash.MatchString(imageHash)
+}
+
+func IsImageValid(image []byte) bool {
+	for _, magicNumber := range magicNumbers {
+		if len(image) < len(magicNumber) {
```
Comment (tobiges) [inline pkg/imageuploader/imageuploader.go:118]: this should be in the isMagicNumberValid func

File: pkg/imageuploader/imageuploader.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+	}
+
+	err = utils.CreateDirIfNotExists(params.ImageDirectory)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ImageUploader{
+		imageDirectory: params.ImageDirectory,
+		maxImageSize:   int64(params.MaxImageSizeMB) * 1024 * 1024,
```
Comment (tobiges) [inline pkg/imageuploader/imageuploader.go:56]: why is this an int64, it is casted to an int in the function below

**Review by tobiges — DISMISSED** [review]

File: pkg/imageuploader/imageuploader.go (hunk 6 above)
Comment (StealWonders) [reply pkg/imageuploader/imageuploader.go:56]: Resolution != File Size

File: internal/graphql/resolvers/helper.go (hunk 1 above)
Comment (StealWonders) [reply internal/graphql/resolvers/helper.go:14]: Should be covered by gqlgen, we should recheck some time

**Review by tobiges — APPROVED** [review]

**Pull request merged**